lib/db, cmd: Add abstraction level to database (ref #4085)

imsodin opened the pull request:
> I think the DB should be an interface, which java world can implement.
https://github.com/syncthing/syncthing/pull/5805#issuecomment-507815753

I propose this interface as an intermediate step to the lib/syncthing PR (#5805) (from [backend.go](https://github.com/syncthing/syncthing/pull/5830/files#diff-ae0519820d68bd6a365eaf4ddcb6c2e6)): 

```golang
var (
	ErrNotFound = errors.New("given key not found in db")
	ErrClosed   = errors.New("db backend was closed")
)

// Backend is the interface that a database needs to implement to be used with
// Syncthing.
// It is not guaranteed that after calling Close no more calls are made to other
// methods. All returned errors due to methods being called after Close must be
// ErrClosed.
// Errors due to a given key not being found in the database, must be ErrNotFound.
type Backend interface {
	Get(key []byte) ([]byte, error)
	Has(key []byte) (bool, error)
	Put(key, val []byte) error
	Delete(key []byte) error
	NewBatch() Batch
	NewIterator(slice *util.Range) (iterator.Iterator, error)
	GetSnapshot() (Snapshot, error)
	Close()
}

// Batch is a container to check in several changes at once to the database to
// ensure a semantically consistent state.
//
// CheckFlush should check the size of the Batch write and call Flush in
// appropriate intervals (e.g. when the Batch exceeds a certain size).
//
// Flush commits the contents of the Batch to db and calls Reset on it on success.
type Batch interface {
	CheckFlush() error
	Delete(key []byte)
	Flush() error
	Len() int
	Put(key, value []byte)
	Reset()
}

// Snapshot provides a consistent view of the database at the time of creation.
// It must be released after use by calling Release.
type Snapshot interface {
	Get(key []byte) ([]byte, error)
	Has(key []byte) (bool, error)
	NewIterator(*util.Range) (iterator.Iterator, error)
	Release()
}
```

The separate `ErrorChecker` interface is due to read only transactions, which shouldn't have the full `Backend` to prevent writing.

Most of lowlevel.go moved to goleveldb.go as an implementation of `Backend` (gitk nicely diffs both lowlevel.go and goleveldb.go against the old lowlevel.go, github apparently doesn't recognize/care about the similarity). The rest of the changes are attributed to adding `error` to various methods in the interface to get rid of the hack types for closed dbs (`closedIter` and `closedSnap`).

AudriusButkevicius commented:
Actually rethinking my own words, I wonder if its likely that Java will implement its own database.

What was the problem with gomobile in general? Does this help solve it? Perhaps turning LowLevel into an interface solves this too and we rid of one of the abstractions?

imsodin commented:
The distinction between `Lowlevel` and `instance` is indeed not necessary - I removed that.

Lowlevel holds some state which has nothing to do with the db backend, so that wouldn't make sense to include in that interface. In general this is not directly relevant to the gomobile problem, it's just that with a user of the library could either choose to use our existing implementation of the `Backend` interface through goleveldb or implement it itself. Having `*LowLevel` in the library interface seems pretty obscure and any user definitely needs to just use our implementation. So in essence in my opinion it's either something like the interface in this PR or then we can just put `dbPath string` into the interface.

AudriusButkevicius commented on lib/db/goleveldb.go:
```diff
@@ -0,0 +1,289 @@
+// OpenGoleveldb attempts to open the database at the given location, and runs
+// recovery on it if opening fails. Worst case, if recovery is not possible,
+// the database is erased and created from scratch.
+func OpenGoleveldb(location string) (Backend, error) {
```
Also, why does this need to be public? For util binaries?

AudriusButkevicius commented on lib/db/goleveldb.go:
```diff
@@ -0,0 +1,289 @@
+
+func (db *goleveldb) IsClosed(err error) bool {
+	return err == leveldb.ErrClosed
+}
```
I dislike this. net package returns net.Error that has IsTimeout() etc, we should have db.Error,

or db.ErrNotFound (oppose to some checker magic).

AudriusButkevicius commented on lib/db/lowlevel.go:
```diff
@@ -99,208 +104,638 @@ func (db *Lowlevel) Committed() int64 {
+	t, err := db.newReadOnlyTransaction()
+	if err != nil {
+		l.Debugln("newReadOnlyTransaction:", err)
+		return
```
We should not be swallowing these silently.
Throughout the file everywhere.

AudriusButkevicius commented on lib/db/lowlevel.go:
```diff
@@ -99,208 +104,638 @@ func (db *Lowlevel) Committed() int64 {
+	t, err := db.newReadWriteTransaction()
+	if err != nil {
+		l.Debugln("newReadWriteTransaction:", err)
+		return
```
bad stuff to hide this

AudriusButkevicius commented on lib/db/lowlevel.go:
```diff
@@ -99,208 +104,638 @@ func (db *Lowlevel) Committed() int64 {
+		f, err := unmarshalTrunc(dbi.Value(), truncate)
+		if err != nil {
+			l.Debugln("unmarshal error:", err)
+			continue
```
again, bad stuff

AudriusButkevicius commented on lib/db/lowlevel.go:
```diff
@@ -99,208 +104,638 @@ func (db *Lowlevel) Committed() int64 {
+// Local flags or the invalid bit might change without the version
+// being bumped. The IsInvalid() method handles both.
+func unchanged(nf, ef FileIntf) bool {
+	return ef.FileVersion().Equal(nf.FileVersion()) && ef.IsInvalid() == nf.IsInvalid()
```
The diff of this file is unreadable.
Idk, perhaps make it over several commits or something to make it possible to understand what's yours and what's just moving.

The whole "assume no error" looks like a disasster to me.

AudriusButkevicius commented on lib/db/namespaced.go:
```diff
@@ -36,28 +36,32 @@ func NewNamespacedKV(db *Lowlevel, prefix string) *NamespacedKV {
+	it, err := n.db.NewIterator(util.BytesPrefix(n.prefix))
+	if err != nil {
+		l.Debugln("NewIterator:", err)
+		return
```
BAAAAD

AudriusButkevicius commented on lib/db/namespaced.go:
```diff
@@ -36,28 +36,32 @@ func NewNamespacedKV(db *Lowlevel, prefix string) *NamespacedKV {
 	var valBs [8]byte
 	binary.BigEndian.PutUint64(valBs[:], uint64(val))
-	n.db.Put(n.prefixedKey(key), valBs[:], nil)
+	n.db.Put(n.prefixedKey(key), valBs[:])
```
Let's fix this, we're making a library, failing silently is bad.

AudriusButkevicius commented on lib/db/schemaupdater.go:
```diff
@@ -91,10 +91,18 @@ func (db *schemaUpdater) updateSchema() error {
+	t, err := db.newReadWriteTransaction()
+	if err != nil {
+		l.Debugln("newReadWriteTransaction:", err)
+		return
```
BAAD

AudriusButkevicius commented on lib/db/smallindex.go:
```diff
@@ -42,7 +41,10 @@ func newSmallIndex(db *leveldb.DB, prefix []byte) *smallIndex {
-	it := i.db.NewIterator(util.BytesPrefix(i.prefix), nil)
+	it, err := i.db.NewIterator(util.BytesPrefix(i.prefix))
+	if err != nil {
+		return
```
bad

AudriusButkevicius commented on lib/db/transactions.go:
```diff
@@ -83,18 +88,22 @@ func (t readOnlyTransaction) getGlobal(keyBuf, folder, file []byte, truncate boo
 func (t readWriteTransaction) close() {
-	t.flush()
+	_ = t.Flush()
 	t.readOnlyTransaction.close()
```
This could return an error, and tell us that it fucked up and didn't do anything so we can let the user know?

imsodin commented:
@AudriusButkevicius I reworked the commits to be more readable and incorporated some of your feedback. I have not done anything about the error handling. I agree that it is/was abysmal and needs improvement. I just first need to think about what we actually want to do: As any error that's not not found/closed is probably fatal, at the moment we either ignore it or panic. I am unsure what the most useful thing would be (on a high level): Shut syncthing down? Display a warning, archive the db and start over on a new one?

imsodin commented on lib/db/lowlevel.go:
```diff
@@ -7,88 +7,93 @@
+// appropriate intervals (e.g. when the Batch exceeds a certain size).
+//
+// Flush commits the contents of the Batch to db and calls Reset on it on success.
+type Batch interface {
```
Im not much of a fan of the "er" suffix, it seems pointless to me. It's the general type that defines how a batch has to behave, and implementations are then `mysqlBatch`, `leveldbBatch` and so on. If any suffix I'd like "Intf" much better. Just imagine "Backender", "Modeller", "Snapshotter", ... urgh

imsodin commented on lib/db/lowlevel.go:
```diff
@@ -7,88 +7,93 @@
+//
+// Flush commits the contents of the Batch to db and calls Reset on it on success.
+type Batch interface {
+	CheckFlush() error
```
My reasoning is, that it can depend on the db when a good size to flush is reached. Letting the db flush on `Put` if necessary would invalidate the purpose of `Batch`: Collecting changes together that should be done atomically (semantically). I.e. `CheckFlush` is caller saying to the db, that it may flush at this point if the db thinks it's a good idea, and `Flush` says do it now (likely because operation ended and nothing more is being added to Batch).

I could create an `flushBatch` type used by `Interface` that takes care of flushing based on size regardless of the underlying `Backend`, then `CheckFlush` could be removed from the interface.

imsodin commented on lib/db/lowlevel.go:
```diff
@@ -7,88 +7,93 @@
+	Put(key, val []byte) error
+	Delete(key []byte) error
+	NewBatch() Batch
+	NewIterator(slice *util.Range) (iterator.Iterator, error)
```
I thought as this are packages separate from the meat of the golevel implementation it could be fine to use them, but sure makes sense to have the full interface/types here. As it is plainly at least inspired by syndtr/goleveldb, I probably need to attribute that somehow? Maybe a comment at the top referencing to `LICENSE-syndtr` which contains the copyright (BSD 2-clause, which is fine with MPL I think). Or am I overthinking this?

imsodin commented:
I have looked a bit into it, and bubbling all errors up to the exported methods in lib/db alone is already huge. Then there is a few places in model where those methods are used: Some in synchronous services (e.g. folders) and some asynchronously (e.g. remote indexes). Any solution that wants to "react moderately" to db errors (e.g. restart folder or connection) would need some kind of error tracking to not retry indefinitely and user facing reporting - and even then there's some danger that some effect of a db failure has been misjudged and in worst case even messes up remotes. Therefore I'd tend to just shut down. It should hardly ever happen and whatever controls Syncthing should already take care of reporting the problem and throttling restarts. And if the db is toast startup will fail, if it isn't it is the cleanest possible restart (index exchange etc). It would also have the added benefit that we don't need to bubble errors all the way up: The database could just get an interface to a "Syncthing controller" (like the rest API already has) and trigger exiting on error (after doing whatever error reporting/printing that is suitable).  
What are your thoughts on this?

AudriusButkevicius commented:
So I do think we should bubble the errors all the way up if we assume this behaves like a library. I don't know how we can do this but I am not a fan of the inverse flow control where we give everyone a "shoot myself" switch.

@calmh what do you think about the error handling issue in general?

calmh commented:
I didn't read this at all yet because I've been on a business trip this week and have had no time/energy. What are we trying to do, make it easier to switch out the lowest level database? Is that ever going to be a thing?

But if it were, I do think we should properly propagate errors upwards, to the boundary of the db package. It's not the db package's responsibility to decide what an error means.

Then at some point in the stack the reality becomes that we cannot operate without the database is working order so we're going to crash when an error is returned. I think it's fine to move that decision as high up as possible though, potentially into the model layer. There could, at that point, exist a type crashingDatabase that simply wraps the database layers, removing error returns by intercepting them and shutting down when an error happens...

imsodin commented:
@calmh This PR is very low priority, it's just a consequence of the idea to have an interface in the proposed lib/syncthing interface for the database. Yes, in principle such that an alternative lowest level database could be used - pretty hypothetically. I think it's a nice change, but it's also a huge change that's currently just change for changes sake, so I am close to letting it drop :) 

The entire error discussion mostly sparked because the bot started screaming at all the places I ignored errors and errors are ignored already today, which isn't a nice state to be in. Meaning having actual error handling, by bubbling up or whatever, is something to do, regardless of whether in this PR or not.
